src/react/hooks/__tests__/useFragment.test.tsx

```
@@ -304,6 +305,380 @@ describe("useFragment", () => {
+        list: [
+          { __ref: 'Item:{"id":1}' },
+          { __ref: 'Item:{"id":2}' },
+          { __ref: 'Item:{"id":3}' },
+        ],
+      },
+      'Item:{"id":1}': {
+        __typename: "Item",
+        id: 1,
+        text: "first",
+      },
+      'Item:{"id":2}': {
+        __typename: "Item",
+        id: 2,
+        text: "second",
+      },
+      'Item:{"id":3}': {
+        __typename: "Item",
+        id: 3,
+        text: "third",
+      },
+    });
+
+    render(
+      <ApolloProvider client={client}>
+        <List />
+      </ApolloProvider>,
+    );
+
+    function getItemTexts() {
+      return screen.getAllByText(/Item #\d+/).map(el => el.textContent);
+    }
+
+    await waitFor(() => {
+      expect(getItemTexts()).toEqual([
+        "Item #1: first",
+        "Item #2: second",
+        "Item #3: third",
+      ]);
+    });
+
+    expect(renders).toEqual([
+      "list",
+      "item 1",
+      "item 2",
+      "item 3",
+    ]);
+
+    function appendLyToText(id: number) {
+      act(() => {
+        cache.modify({
+          id: cache.identify({ __typename: "Item", id })!,
+          fields: {
+            text(existing) {
+              return existing + "ly";
+            },
+          },
+        });
+      });
+    }
+
+    appendLyToText(2);
+
+    await waitFor(() => {
+      expect(renders).toEqual([
+        "list",
+        "item 1",
+        "item 2",
+        "item 3",
+        "item 2",
+      ]);
+
+      expect(getItemTexts()).toEqual([
+        "Item #1: first",
+        "Item #2: secondly",
+        "Item #3: third",
+      ]);
+    });
+
+    appendLyToText(1);
+
+    await waitFor(() => {
+      expect(renders).toEqual([
+        "list",
+        "item 1",
+        "item 2",
+        "item 3",
+        "item 2",
+        "item 1",
+      ]);
+
+      expect(getItemTexts()).toEqual([
+        "Item #1: firstly",
+        "Item #2: secondly",
+        "Item #3: third",
+      ]);
+    });
```

I see you disabled the `no-wait-for-multiple-assertions` rule. It looks like the inspiration for that rule was so that you don't have to wait for the timeout before you see a test failure on a subsequent expect. See [this doc](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#having-multiple-assertions-in-a-single-waitfor-callback).

A way that I think you can structure this to both adhere to the existing rule and keep this readable is to flip these assertions around: Wait for the UI to update with the expected results, then validate that you have the right amount of renders. This tells a nice story with the test as well (i.e. "when my UI reflects its expected state, its rendered the right amount of times" vs "when its rendered the right amount of times, then my UI looks like X").

Here is how I'd recommend structuring this test to allow us to maintain that rule:

```ts
await waitFor(() => {
  expect(getItemTexts()).toEqual([
    "Item #1: firstly",
    "Item #2: secondly",
    "Item #3: third",
  ]);
});

expect(renders).toEqual([...]);
```

---

I disagree with the wisdom of this rule, because I do not want there to be any asynchronous pause between these checks. As soon as the expected `renders` have been performed, I want to check _right then_ that what was rendered matches my expectations.

The rationale in [that doc](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#having-multiple-assertions-in-a-single-waitfor-callback) talks about "failing faster" which is not relevant here, and claims the separations helps "wait for the UI to settle to the state we want to assert on" which would be worrisome if it were actually necessary here. I don't want to make things work by injecting extra pauses here and there, if we can help it. This works now, and I'd like to keep testing it this way.

---

Your rationale makes sense. Thanks!